inlang/packages/paraglide/paraglide-js/src/compiler/runtime/extract-locale-from-url.js

```diff
@@ -6,38 +6,54 @@ import {
 	urlPatterns,
 } from "./variables.js";
 
+/**
+ * If extractLocaleFromUrl is called many times on the same page and the URL
+ * hasn't changed, we don't need to recompute it every time which can get expensive.
+ */
+const urlToLocaleCache = new Map();
+
 /**
  * Extracts the locale from a given URL using native URLPattern.
  *
  * @param {URL|string} url - The full URL from which to extract the locale.
  * @returns {Locale|undefined} The extracted locale, or undefined if no locale is found.
  */
 export function extractLocaleFromUrl(url) {
-	if (TREE_SHAKE_DEFAULT_URL_PATTERN_USED) {
-		return defaultUrlPatternExtractLocale(url);
+	const urlString = typeof url === "string" ? url : url.href;
+	
+	if (urlToLocaleCache.has(urlString)) {
+		return urlToLocaleCache.get(urlString);
 	}
+	
+	let result;
+	if (TREE_SHAKE_DEFAULT_URL_PATTERN_USED) {
+		result = defaultUrlPqatternExtractLocale(url);
+	} else {
+		const urlObj = typeof url === "string" ? new URL(url) : url;
 
-	const urlObj = typeof url === "string" ? new URL(url) : url;
-
-	// Iterate over URL patterns
-	for (const element of urlPatterns) {
-		for (const [locale, localizedPattern] of element.localized) {
-			const match = new URLPattern(localizedPattern, urlObj.href).exec(
-				urlObj.href
-			);
+		// Iterate over URL patterns
+		for (const element of urlPatterns) {
+			for (const [locale, localizedPattern] of element.localized) {
+				const match = new URLPattern(localizedPattern, urlObj.href).exec(
+					urlObj.href
+				);
 
-			if (!match) {
-				continue;
-			}
+				if (!match) {
+					continue;
+				}
 
-			// Check if the locale is valid
-			if (assertIsLocale(locale)) {
-				return locale;
+				// Check if the locale is valid
+				if (assertIsLocale(locale)) {
+					result = locale;
+					break;
+				}
 			}
+			if (result) break;
 		}
 	}
-
-	return undefined;
+	
+	urlToLocaleCache.set(urlString, result);
```

Comment:
The cache will infinitely grow in size. 

If we do this, we should have some sort of LRU cache. This might be enough: 

```ts
class LRUCache {
  constructor(max = 100) {
    this.max = max;
    this.map = new Map();
  }
  get(key) {
    if (!this.map.has(key)) return undefined;
    const val = this.map.get(key);
    // refresh insertion order
    this.map.delete(key);
    this.map.set(key, val);
    return val;
  }
  set(key, val) {
    if (this.map.has(key)) this.map.delete(key);
    this.map.set(key, val);
    // if we’re over capacity, drop the oldest
    if (this.map.size > this.max) {
      const oldestKey = this.map.keys().next().value;
      this.map.delete(oldestKey);
    }
  }
}

const cache = new LRUCache(200);
```

Reply:
Hmm, maybe to make it even simpler, storing a single pair `url`, `locale` is enough? Or are there scenarios where `extractLocaleFromUrl` is called on a set of repeated distinct URLs?

Reply:
On the server extractLocaleFromUrl can be called for N routes. But given that your problem was client side, starting with a  single pair should fix the client side performance issues

Reply:
On my home page https://esim.info, `extractLocalFromUrl` is called 1500 times, always with the same argument. So caching the result is sensible. I don't think we need to cache more than one URL but I'm open to do it.

Reply:
Let's do the simple implementation. Avoids caching pitfalls

Reply:
I used `console.count` on the server and `extractLocalFromUrl` isn't called as many time (1500 times on the client to load my home page, only 4 times on the server to SSR it)